e2e_tests/tests/cluster/test_ship_logs.py

```diff
@@ -0,0 +1,512 @@
+
+            print("ready!", flush=True)
+
+            time.sleep(5)
+            """
+        )
+        with ShipLogServer() as srv:
+            # Start a subprocess so we can signal it.
+            env = {
+                "DET_MASTER": srv.master_url(),
+                "DET_SESSION_TOKEN": "token",
+                "DET_TASK_LOGGING_METADATA": "{}",
+                "DET_SHIPPER_EMIT_STDOUT_LOGS": "1",
+            }
+            fullcmd = [sys.executable, "-u", ship_logs.__file__] + cmd
+            p = subprocess.Popen(fullcmd, env=env, stdout=subprocess.PIPE)
+            assert p.stdout
+            try:
+                # Wait for the granchild log to indicate signals are set up.
+                for line in p.stdout:
+                    if b"ready!" in line:
+                        break
+                # Send a signal that is caught and logged, to test signal forwarding.
+                p.send_signal(signal.SIGINT)
+                for line in p.stdout:
+                    if b"caught sigint!" in line:
+                        break
+                # Send a signal that is not caught, to test for signal exit codes.
+                p.send_signal(signal.SIGTERM)
+                exit_code = p.wait()
+            finally:
+                p.kill()
+                p.wait()
+        assert exit_code == 128 + signal.SIGTERM, exit_code
+        assert "".join(srv.logs) == "ready!\ncaught sigint!\n", srv.logs
+
+    @pytest.mark.e2e_cpu
+    def test_exit_wait_time(self) -> None:
+        cmd = mkcmd("print('hello world')")
+        # We need a misbehaving server to guarantee the shipper times out.
+        # This misbehaving server will listen without ever accepting.
+        with socket.socket() as listener:
+            listener.bind(("127.0.0.1", 0))
+            listener.listen(10)
+            _, port = listener.getsockname()
+            master_url = f"http://127.0.0.1:{port}"
+            start = time.time()
+            exit_code = self.run_ship_logs(master_url, cmd, log_wait_time=0.1)
+            end = time.time()
+            assert exit_code == 0, exit_code
+            assert end - start < 1, end - start
+
+    @pytest.mark.e2e_cpu
+    def test_entrypoint_not_found(self) -> None:
+        cmd = ["/does-not-exist"]
+        with ShipLogServer() as srv:
+            exit_code = self.run_ship_logs(srv.master_url(), cmd)
+        # 127 is the standard bash exit code for file-not-found.
+        assert exit_code == 127, exit_code
+        assert "FileNotFoundError" in "".join(srv.logs), srv.logs
+
+    @pytest.mark.e2e_cpu
+    def test_entrypoint_not_executable(self) -> None:
+        cmd = ["/bin/"]
+        with ShipLogServer() as srv:
+            exit_code = self.run_ship_logs(srv.master_url(), cmd)
+        # 126 is the standard bash exit code for permission failure.
+        assert exit_code == 126, exit_code
+        assert "PermissionError" in "".join(srv.logs), srv.logs
+
+    @pytest.mark.e2e_cpu
```

Comment:
I like the test, and the internet tells me there's no better way to test whether a library is standard library besides keeping a list, but this implementation feels hacky. Possibly hacky for a good reason? Why didn't you test against a list directly in the main thread?

Reply:
Hacky in what way?

You can't test in the main thread; you need a fresh interpreter because if for instance `pytest` is imported in the main thread already, you can't tell if importing `ship_logs` depends on `pytest` or not.

See also `harness/tests/test_imports.py`, we have a similar approach there (but a blacklist instead of a whitelist, since it is for quite a different purpose).

Reply:
Ugh. I hadn't realized that. That makes sense.

Well... is it less hacky to [get the filename](https://docs.python.org/3/library/importlib.html#importlib.util.find_spec) the import refers to and then reading/parsing that file? Maybe with an [ast.parse](https://docs.python.org/3/library/ast.html#ast.parse)?

Reply:
That feels substantially more hacky to me.

I feel 100% confident in the solution I wrote.

I had an earlier solution which did the import and directly inspected the contents of `sys.modules` but there's some confusing things that can go on there.  I wasn't 100% confident in that solution.

Likewise, I wouldn't be 100% confident in parsing the code because the only bulletproof way to know what a thing imports is to execute it with the real python interpreter.